### PR TITLE
postgres: databroker storage backend

### PR DESCRIPTION
**Pull request by calebdoxsey**
## Summary
Add a new postgres databroker storage backend.

## Related issues
Fixes https://github.com/pomerium/internal/issues/852

## Checklist
- [x] reference any related issues
- [ ] updated docs
- [x] updated unit tests
- [ ] updated UPGRADING.md
- [x] add appropriate tag (`improvement` / `bug` / etc)
- [x] ready for review

### CODE REVIEW

File: pkg/storage/postgres/postgres.go [hunk 1]
```diff
@@ -0,0 +1,322 @@
...
+		return pgtype.JSONB{Status: pgtype.Null}
+	}
+
+	bs, err := protojson.Marshal(any)
+	if err != nil {
+		log.Warn(context.Background()).
+			Err(err).
+			Str("data", prototext.Format(any)).
+			Msg("storage/postgres: error marshaling any to JSON")
+		return pgtype.JSONB{Status: pgtype.Null}
```
Comment (wasaga) [inline pkg/storage/postgres/postgres.go:300]: shouldn't we bubble that error up and fail record update?

File: pkg/storage/postgres/stream.go [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+		}
+
+		if !block {
+			return false
+		}
+
+		select {
+		case <-stream.ctx.Done():
+			stream.err = stream.ctx.Err()
+			return false
```
Comment (wasaga) [inline pkg/storage/postgres/stream.go:152]: shouldn't we stop the ticker here just in case

**Review by wasaga — COMMENTED** [review]
can we also add tests for `filter.go` ?

File: internal/identity/oauth/github/github.go [hunk 3]
```diff
@@ -42,7 +42,7 @@ const (
 	refreshDeadline = time.Minute * 60
 )
 
-var maxTime = time.Unix(1<<63-1, 0)
+var maxTime = time.Unix(253370793661, 0) // year 9999
```
[the remark this answers is not in the document]
Comment (calebdoxsey) [reply internal/identity/oauth/github/github.go:45]: It is somewhat related. The previous timestamp we were using was beyond the valid range for a protobuf timestamp which then caused problems when we called `protojson.Marshal`. This new year is a valid `protojson` time value, so we can store the data in Postgres.

File: internal/databroker/registry.go [hunk 4]
```diff
@@ -90,7 +90,7 @@ func (srv *Server) newRegistryLocked() (registry.Interface, error) {
 	ctx := context.Background()
 
 	switch srv.cfg.storageType {
-	case config.StorageInMemoryName:
+	case config.StorageInMemoryName, config.StoragePostgresName:
```
[the remark this answers is not in the document]
Comment (calebdoxsey) [reply internal/databroker/registry.go:93]: Registry support was not part of the requirements. I don't know what's involved in implementing that.

File: pkg/storage/postgres/stream.go (hunk 2 above)
Comment (calebdoxsey) [reply pkg/storage/postgres/stream.go:152]: The ticker is stopped in Close.

File: internal/databroker/registry.go [hunk 5]
```diff
@@ -90,7 +90,7 @@ func (srv *Server) newRegistryLocked() (registry.Interface, error) {
...
 		log.Info(ctx).Msg("using in-memory registry")
```
[the remark this answers is not in the document]
Comment (calebdoxsey) [reply internal/databroker/registry.go:94]: It is accurate.

File: internal/databroker/registry.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (calebdoxsey) [reply internal/databroker/registry.go:93]: I did not implement the registry for Postgres. It was not part of the requirements. I can work on this if we want.

**Review by desimone — COMMENTED** [review]
Looks like we need to carve out an exception for code-coverage, but otherwise LGTM!

**Review by wasaga — APPROVED** [review]

**Pull request merged**